website/blog/2024-07-26-pyfunc-in-practice/index.md

```diff
@@ -0,0 +1,930 @@
+
+Now that everything has been encapsulated in a PyFunc model, the lifecycle of the ensemble model closely mirrors that of a traditional MLflow model. The following diagram depicts the lifecycle of the model.
+
+![Ensemble Model Lifecycle](ensemble-model-lifecycle.png)
+
+
+## MLflow Tracking
+### Using the `fit` Method to Train Sub-Models  
+
+Once the data is preprocessed, we use the custom `fit` method to train all the sub-models in our Ensemble Model. This method applies grid search to find the best hyperparameters for each sub-model, fits them to the training data, and saves the trained models for future use.
+
+```python
+# Initialize the MLflow client
+client = mlflow.MlflowClient()
+
+# Set the experiment name for organizing runs in MLflow
+mlflow.set_experiment("Ensemble Model")
+
+# Load dataset from the provided URL
+data = pd.read_csv(
+    "https://github.com/zobi123/Machine-Learning-project-with-Python/blob/master/Housing.csv?raw=true"
+)
+
+# Separate features and target variable
+X = data.drop("price", axis=1)
+y = data["price"]
+
+# Split dataset into training and test sets
+X_train, X_test, y_train, y_test = train_test_split(
+    X, y, test_size=0.2, random_state=42
+)
+
+# Create a directory to save the models and related files
+os.makedirs("models", exist_ok=True)
+
+# Initialize and train the EnsembleModel
+ensemble_model = EnsembleModel()
+
+# Preprocess the training data using the defined feature engineering method
+X_train_processed = ensemble_model.feature_engineering(X_train)
+
+# Fit the models with the preprocessed training data and save them
+ensemble_model.fit(X_train_processed, y_train, save_path="models")
+
+# Infer the model signature using a small example from the training data
+example_input = X_train[:1]  # Use a single sample for signature inference
+example_input["strategy"] = "average"
+example_output = y_train[:1]
+signature = infer_signature(example_input, example_output)
+
+# Save the preprocessing pipeline to disk
+joblib.dump(ensemble_model.preprocessor, "models/preprocessor.pkl")
+
+# Define strategies for the ensemble model
+strategy_data = {
+    "strategy": [
+        "average_1",
+        "average_2",
+        "weighted_1",
+        "weighted_2",
+        "weighted_3",
+        "weighted_4",
+    ],
+    "model_list": [
+        "random_forest,xgboost,decision_tree,gradient_boosting,adaboost",
+        "decision_tree",
+        "random_forest,xgboost,decision_tree,gradient_boosting,adaboost",
+        "random_forest,xgboost,gradient_boosting",
+        "decision_tree,adaboost",
+        "xgboost,gradient_boosting",
+    ],
+    "weights": ["1", "1", "0.2,0.3,0.1,0.2,0.2", "0.4,0.4,0.2", "0.5,0.5", "0.7,0.3"],
+}
+
+# Create a DataFrame to hold the strategy information
+strategies_df = pd.DataFrame(strategy_data)
+
+# Add strategies to the database
+ensemble_model.add_strategy_and_save_to_db(strategies_df, "models/strategies.db")
+
+# Define the Conda environment configuration for the MLflow model
```

Comment:
Is this being done purely for the requirements? Does defining the `requirements` argument in the `log_model` call not sufficient?

Reply:
You're totally right; defining the Conda environment as a dictionary directly in the `log_model` call is more efficient, there's no need to use the file option. Updated.